Pal/lib/memmgr.h

```diff
@@ -216,10 +230,16 @@ static inline OBJ_TYPE * get_mem_obj_from_mgr_enlarge (MEM_MGR mgr,
         system_lock();
         area->size = size;
         INIT_LIST_HEAD(area, __list);
+
+        /* There can be concurrent operations to extend the manager. In case
+         * someone has already enlarge the space, we just add the new area to
+         * the list for later use. */
         listp_add(area, &mgr->area_list, __list);
-        __set_free_mem_area(area, mgr, size);
+        if (mgr_size == mgr->size) /* check if the size has changed */
+            __set_free_mem_area(area, mgr);
```

Comment:
`area` leaks if the condition of the `if` isn't satisfied.

Reply:
Hmm, or maybe do you allow the case where it's only added to the `area_list`? IMO it may drain memory in some unlucky scenarios.